### PR TITLE
Add `clone_from` specializations to `Rc` and `Arc`

### PR DESCRIPTION
**Pull request by Lucretiel**
Add specializations for `clone_from` to `Rc` and `Arc`. These check to see if the pointers already share ownership of the same object and, if so, don't do any additional work.

### CODE REVIEW

File: library/alloc/src/rc.rs [hunk 1]
```diff
@@ -1627,6 +1627,15 @@ impl<T: ?Sized> Clone for Rc<T> {
             Self::from_inner(self.ptr)
         }
     }
+
+    #[inline]
+    fn clone_from(&mut self, source: &Self) {
+        // If `self` and `source` share ownership of the same value, no extra
+        // work needs to happen
+        if self.ptr != other.ptr {
```
Comment (eggyal) [inline library/alloc/src/rc.rs:1635]:
```suggestion
        if self.ptr != source.ptr {
```

File: library/alloc/src/rc.rs [hunk 2]
```diff
@@ -2558,6 +2567,15 @@ impl<T: ?Sized> Clone for Weak<T> {
         }
         Weak { ptr: self.ptr }
     }
+
+    #[inline]
+    fn clone_from(&mut self, source: &Self) {
+        // If `self` and `source` share the same resource, nothing else needs
+        // to happen
+        if self.ptr != other.ptr {
```
Comment (eggyal) [inline library/alloc/src/rc.rs:2575]:
```suggestion
        if self.ptr != source.ptr {
```

File: library/alloc/src/sync.rs [hunk 3]
```diff
@@ -1525,6 +1525,15 @@ impl<T: ?Sized> Clone for Arc<T> {
 
         unsafe { Self::from_inner(self.ptr) }
     }
+
+    #[inline]
+    fn clone_from(&mut self, source: &Self) {
+        // If `self` and `source` share the same resource, nothing else needs
+        // to happen
+        if self.ptr != other.ptr {
```
Comment (eggyal) [inline library/alloc/src/sync.rs:1533]:
```suggestion
        if self.ptr != source.ptr {
```

File: library/alloc/src/sync.rs [hunk 4]
```diff
@@ -2330,6 +2339,15 @@ impl<T: ?Sized> Clone for Weak<T> {
 
         Weak { ptr: self.ptr }
     }
+
+    #[inline]
+    fn clone_from(&mut self, source: &Self) {
+        // If `self` and `source` share the same resource, nothing else needs
+        // to happen
+        if self.ptr != other.ptr {
```
Comment (eggyal) [inline library/alloc/src/sync.rs:2347]:
```suggestion
        if self.ptr != source.ptr {
```

**Review by eggyal — CHANGES_REQUESTED** [review]
Just a typo repeated in each implementation, but otherwise seems sensible to me.

**Review by eggyal — APPROVED** [review]

**Comment by mejrs** [comment]
Do you have a benchmark for this?

**Comment by Lucretiel** [comment]
Not yet; I can write one this weekend. Are there any existing `Arc` benchmarks I can base off of?

**Comment by mejrs** [comment]
I have no idea, sorry.

**Comment by the8472** [comment]
I think there's another case we can optimize. If the refcount of `self` is exactly 1 and `T: Clone` then we can call `self.get_mut().clone_from(source)`

**Comment by Lucretiel** [comment]
> I think there's another case we can optimize. If the refcount of `self` is exactly 1 and `T: Clone` then we can call `self.get_mut().clone_from(source)`

I'm not sure that's the case; `Rc::clone_from` and `Arc::clone_from` carry with them the requirement that the original object is now shares ownership with the source.

**Comment by the8472** [comment]
Ah right, but then I wonder how much this actually helps. It'll only avoid an atomic increment, in a quite narrow scenario.
Benchmarks would help but even then the question is does it occur often enough in the wild that it's worth it?

**Comment by Dylan-DPC** [comment]
@Lucretiel any updates on this? thanks?

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**